File: VRTVungleManager.m
//
//  VRTIronSourceManager.m
//  VrtcalSDKInternalTestApp
//
//  Created by Scott McCoy on 12/20/21.
//  Copyright © 2021 VRTCAL. All rights reserved.
//

//Header
#import "VRTVungleManager.h"

//Dependencies
#import <VrtcalSDK/VrtcalSDK.h>
#import <VungleSDK/VungleSDK.h>

@interface VRTVungleManager() <VungleSDKDelegate>
@property VungleSDK *sdk;
@property(nonatomic) NSMapTable<NSString*, id<VRTVungleManagerDelegate>> *vrtVungleManagerDelegates;
@end

@implementation VRTVungleManager

+ (instancetype)singleton {
    static VRTVungleManager *singleton = nil;
    static dispatch_once_t onceToken;

    dispatch_once(&onceToken, ^{
        singleton = [[self alloc] init];
    });
    return singleton;
}

- (instancetype)init {
    self = [super init];

    self.sdk = [VungleSDK sharedSDK];
    self.sdk.delegate = self;
    
    //Init the map table such that it will not retain adapters referenced by it
    self.vrtVungleManagerDelegates = [NSMapTable mapTableWithKeyOptions:NSPointerFunctionsStrongMemory valueOptions:NSPointerFunctionsWeakMemory];
    
    
    return self;
}

#pragma mark - API
-(NSError* _Nullable) loadPlacementWithID:(nonnull NSString *)placementID
                    withSize:(VungleAdSize)size
           vrtVungleManagerDelegate:(nonnull id <VRTVungleManagerDelegate>) vrtVungleManagerDelegate {
    

    [self addDelegate:vrtVungleManagerDelegate forPlacementID:placementID];

    
    NSError *error = nil;
    [self.sdk
        loadPlacementWithID:(NSString *)placementID
        withSize:(VungleAdSize)size
        error:&error
    ];
    
    return error;
}

-(nullable NSError*) showBanner:(nonnull NSString*)placementId inContainerView:(UIView*)containerView {
    NSError* error = nil;
    [self.sdk
        addAdViewToView:containerView
        withOptions:nil
        placementID:placementId
        error:&error
    ];
    
    return error;
}

- (nullable NSError*) showInterstitial:(NSString*)placementId viewController:(UIViewController*)viewController {
    NSError* error = nil;
    
    [self.sdk
        playAd:viewController
        options:nil
        placementID:placementId
        error:&error
    ];
    
    return error;
}


#pragma mark - Delegate Pass-Through
- (void)vungleWillShowAdForPlacementID:(nullable NSString *)placementID {
    [[self delegateForPlacementId:placementID] vungleWillShowAdForPlacementID:placementID];
}

- (void)vungleDidShowAdForPlacementID:(nullable NSString *)placementID {
    [[self delegateForPlacementId:placementID] vungleDidShowAdForPlacementID:placementID];
}

- (void)vungleAdViewedForPlacement:(NSString *)placementID {
    [[self delegateForPlacementId:placementID] vungleAdViewedForPlacement:placementID];
}

- (void)vungleWillCloseAdForPlacementID:(nonnull NSString *)placementID {
    [[self delegateForPlacementId:placementID] vungleWillCloseAdForPlacementID:placementID];
}

- (void)vungleDidCloseAdForPlacementID:(nonnull NSString *)placementID {
    [[self delegateForPlacementId:placementID] vungleDidCloseAdForPlacementID:placementID];
}

- (void)vungleTrackClickForPlacementID:(nullable NSString *)placementID {
    [[self delegateForPlacementId:placementID] vungleTrackClickForPlacementID:placementID];
}

- (void)vungleWillLeaveApplicationForPlacementID:(nullable NSString *)placementID {
    [[self delegateForPlacementId:placementID] vungleWillLeaveApplicationForPlacementID:placementID];
}

- (void)vungleRewardUserForPlacementID:(nullable NSString *)placementID {
    [[self delegateForPlacementId:placementID] vungleRewardUserForPlacementID:placementID];
}

- (void)vungleAdPlayabilityUpdate:(BOOL)isAdPlayable placementID:(nullable NSString *)placementID error:(nullable NSError *)error {
    [[self delegateForPlacementId:placementID] vungleAdPlayabilityUpdate:isAdPlayable placementID:placementID error:error];
}


#pragma mark - NSMapTable Facade
- (void)addDelegate:(id<VRTVungleManagerDelegate>)vrtVungleManagerDelegate forPlacementID:(NSString *)placementId {
    @synchronized(self) {
        [self.vrtVungleManagerDelegates setObject:vrtVungleManagerDelegate forKey:placementId];
    }
}

- (id<VRTVungleManagerDelegate>) delegateForPlacementId:(NSString *)placementId {
    id<VRTVungleManagerDelegate> delegate;
    @synchronized(self) {
        delegate = [self.vrtVungleManagerDelegates objectForKey:placementId];
    }
    return delegate;
}

- (void)removeDelegateForPlacementId:(NSString *)placementId {
    @synchronized(self) {
        [self.vrtVungleManagerDelegates removeObjectForKey:placementId];
    }
}

@end

File: VRTVungleManager.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.26605 - https://swiftify.com/
//
//  VRTVungleManager.swift
//  VrtcalSDKInternalTestApp
//
//  Created by Scott McCoy on 12/20/21.
//  Copyright © 2021 VRTCAL. All rights reserved.
//
//
//  VRTIronSourceManager.m
//  VrtcalSDKInternalTestApp
//
//  Created by Scott McCoy on 12/20/21.
//  Copyright © 2021 VRTCAL. All rights reserved.
//

//Header
import Foundation
import UIKit
import VungleSDK

class VRTVungleManager: NSObject, VungleSDKDelegate {
    private var sdk: VungleSDK?
    private var vrtVungleManagerDelegates: NSMapTable<String, VRTVungleManagerDelegate?>?

    static let singletonVar = self.init()

    class func singleton() -> Self {

        // [Swiftify] `dispatch_once()` call was converted to the initializer of the `singletonVar` variable
        return singletonVar
    }

    override init() {
        super.init()

        sdk = .shared()
        sdk?.delegate = self

        //Init the map table such that it will not retain adapters referenced by it
        vrtVungleManagerDelegates = NSMapTable(keyOptions: .strongMemory, valueOptions: .weakMemory)


    }

    // MARK: - API

    func loadPlacement(
        withID placementID: String,
        with size: VungleAdSize,
        vrtVungleManagerDelegate: VRTVungleManagerDelegate
    ) -> Error? {


        add(vrtVungleManagerDelegate, forPlacementID: placementID)


        var error: Error?
        sdk?.loadPlacement(
            withID: placementID,
            with: size,
            error: &error)

        return error
    }

    func showBanner(_ placementId: String, inContainerView containerView: UIView) -> Error? {
        var error: Error?
        sdk?.addAdView(
            to: containerView,
            withOptions: nil,
            placementID: placementId,
            error: &error)

        return error
    }

    func showInterstitial(_ placementId: String, viewController: UIViewController) -> Error? {
        var error: Error?

        sdk?.playAd(
            viewController,
            options: nil,
            placementID: placementId,
            error: &error)

        return error
    }

    // MARK: - Delegate Pass-Through

    func vungleWillShowAd(forPlacementID placementID: String?) {
        delegate(forPlacementId: placementID)?.vungleWillShowAd(forPlacementID: placementID)
    }

    func vungleDidShowAd(forPlacementID placementID: String?) {
        delegate(forPlacementId: placementID)?.vungleDidShowAd(forPlacementID: placementID)
    }

    func vungleAdViewed(forPlacement placementID: String?) {
        delegate(forPlacementId: placementID)?.vungleAdViewed(forPlacement: placementID)
    }

    func vungleWillCloseAd(forPlacementID placementID: String) {
        delegate(forPlacementId: placementID)?.vungleWillCloseAd(forPlacementID: placementID)
    }

    func vungleDidCloseAd(forPlacementID placementID: String) {
        delegate(forPlacementId: placementID)?.vungleDidCloseAd(forPlacementID: placementID)
    }

    func vungleTrackClick(forPlacementID placementID: String?) {
        delegate(forPlacementId: placementID)?.vungleTrackClick(forPlacementID: placementID)
    }

    func vungleWillLeaveApplication(forPlacementID placementID: String?) {
        delegate(forPlacementId: placementID)?.vungleWillLeaveApplication(forPlacementID: placementID)
    }

    func vungleRewardUser(forPlacementID placementID: String?) {
        delegate(forPlacementId: placementID)?.vungleRewardUser(forPlacementID: placementID)
    }

    func vungleAdPlayabilityUpdate(_ isAdPlayable: Bool, placementID: String?, error: Error?) {
        delegate(forPlacementId: placementID)?.vungleAdPlayabilityUpdate(isAdPlayable, placementID: placementID, error: error)
    }

    // MARK: - NSMapTable Facade

    func add(_ vrtVungleManagerDelegate: VRTVungleManagerDelegate?, forPlacementID placementId: String?) {
        objc_sync_enter(self)
        vrtVungleManagerDelegates?.set(vrtVungleManagerDelegate, forKey: placementId ?? "")
        objc_sync_exit(self)
    }

    func delegate(forPlacementId placementId: String?) -> VRTVungleManagerDelegate? {
        var delegate: VRTVungleManagerDelegate?
        objc_sync_enter(self)
        delegate = vrtVungleManagerDelegates?[placementId ?? ""] as? VRTVungleManagerDelegate
        objc_sync_exit(self)
        return delegate
    }

    func removeDelegate(forPlacementId placementId: String?) {
        objc_sync_enter(self)
        vrtVungleManagerDelegates?.removeObject(forKey: placementId)
        objc_sync_exit(self)
    }
}

//Dependencies